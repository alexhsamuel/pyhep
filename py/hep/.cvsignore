*.pyc
*.pyo
config.py
*-install
pdt.pickle
