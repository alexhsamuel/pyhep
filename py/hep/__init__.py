#-----------------------------------------------------------------------
#
# module hep
#
# Copyright 2003 by Alex Samuel.  All rights reserved.
#
#-----------------------------------------------------------------------

"""Module for general-purpose code."""

#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

from __future__ import generators

import os
import signal
import traceback

#-----------------------------------------------------------------------
# exceptions
#-----------------------------------------------------------------------

class TimeoutException(Exception):
    """A timeout occurred."""

    pass



class NotSupportedError(RuntimeError):
    """A feature is not supported."""

    pass



class SyntaxError(Exception):

    pass



#-----------------------------------------------------------------------
# classes
#-----------------------------------------------------------------------

class Token:

    pass



class Timeout:

    def __timeout_handler(self, signal_number, frame):
        assert signal_number == signal.SIGALRM
        signal.signal(signal.SIGALRM, self.__old_handler)
        self.__active = 0
        raise TimeoutException, self.__message


    def __init__(self, time, message):
        self.__message = message
        self.__old_handler = \
            signal.signal(signal.SIGALRM, self.__timeout_handler)
        self.__active = 1
        signal.alarm(time)


    def cancel(self):
        if not self.__active:
            return
        signal.alarm(0)
        signal.signal(signal.SIGALRM, signal.SIG_DFL)
        self.__active = 0



class MappingAsObject:
    """Implement the object attribute protocol using a mapping.

    A 'MappingAsObject' wraps an object satifying the mapping protocol.
    Attribute references are converted to item gets and sets in the
    underlying mapping."""

    def __init__(self, mapping):
        """Wrap 'mapping' in the object procotol."""
        
        self.__dict__["_MappingAsObject__mapping"] = mapping


    _mapping = property(lambda self: self.__mapping)
    """Read-only access to the underlying mapping."""


    def __getattr__(self, name):
        try:
            return self.__mapping[name]
        except KeyError:
            raise AttributeError, name


    def __setattr__(self, name, value):
        try:
            self.__mapping[name] = value
        except KeyError, exception:
            raise AttributeError, str(exception)



class NrBreitWigner:
    """A non-resonant Breit-Wigner line shape amplitude factor."""

    def __init__(self, mass, width):
        """Construct a non-relativistic Breit-Wigner function.

        'mass' -- The resonance mass.
        
        'width' -- The resonance width."""

        self.__mass = mass
        self.__width = width


    def __call__(self, s):
        """Return the amplitude at energy-squared 's'."""

        mass = self.__mass
        return 1 / (s - complex(mass * mass, mass * self.__width))



#-----------------------------------------------------------------------
# functions
#-----------------------------------------------------------------------

try:
    enumerate = __builtins__.enumerate
except AttributeError:
    def enumerate(iterable):
        count = 0
        for item in iter(iterable):
            yield count, item
            count += 1


try:
    sorted = __builtins__.sorted
except AttributeError:
    def sorted(iterable, cmp=cmp):
        result = list(iterable)
        result.sort(cmp)
        return result


try:
    reversed = __builtins__.reversed
except AttributeError:
    def reversed(iterable):
        result = list(iterable)
        result.reverse()
        return result
    

def sort(sequence, sort_fn=None):
    """Return a copy of 'sequence', sorted by 'sort_fn'.

    'sort_fn' -- The comparison function to use.  If 'None', the default
    (built-in 'cmp') is used."""

    sequence = list(sequence)
    if sort_fn is None:
        sort_fn = cmp
    sequence.sort(sort_fn)
    return sequence


def binarySearch(sequence, value, compare=cmp):
    """Find 'value' in sorted 'sequence' by binary search.

    'sequence' -- A sequence of values.  It is assumed to be sorted by
    'compare'.

    'compare' -- A comparison function.

    returns -- The index of the largest element in 'sequence' smaller
    than 'value', or -1 if there is no such element."""

    if len(sequence) == 0:
        return -1
    if compare(value, sequence[0]) < 0:
        return -1
    if compare(value, sequence[-1]) > 0:
        return len(squence) - 1

    i0 = 0
    i2 = len(sequence) - 1
    while True:
        if i2 <= i0 + 1:
            return i2 - 1
        i1 = (i0 + i2) // 2
        if compare(value, sequence[i1]) < 0:
            i2 = i1
        else:
            i0 = i1


def mapFromKeys(key_sequence, value_function):
    """Construct a map from a sequence of keys.

    A map is created whose keys are the elements of 'key_sequence'.  For
    each key, the corresponding value is constructed from the key using
    'value_function'.  If 'sequence' contains a particular key more than
    once, only one is placed in the map (which one is undefined).

    'key_sequence' -- A sequence of key values.

    'value_function' -- A callable for determining the value for each
    element of the sequence.  It is invoked for each key, passing the
    kay as the single argument."""
    
    result = {}
    for key in key_sequence:
        result[key] = value_function(key)
    return result


def mapFromValues(key_function, value_sequence):
    """Construct a map from a sequence of values.

    A map is created whose values are the elements of 'value_sequence'.
    For each value, the corresponding key is constructed from the value
    using 'key_function'.  If 'key_function' assigns the same key to
    more than one element of the sequence, only one element is actually
    used in the map (which one is undefined).

    'key_function' -- A callable for determining the key for each
    element of the sequence.  It is invoked for each value, passing
    the value as the single argument.

    'value_sequence' -- A sequence of values."""
    
    result = {}
    for value in value_sequence:
        result[key_function(value)] = value
    return result


def invertMap(map):
    """Produce a dictionary by inverting keys and values of 'map'."""

    return dict([ (value, key) for key, value in map.items() ])


def remove(dict, key):
    """Get and remove a dictionary item.

    'dict' -- The dictionary from which to get and remove the item.

    'key' -- The key of the item to get and remove.

    returns -- The value corresponding to 'key' in 'dict'.  

    raises -- 'KeyError' if 'dict' has no key 'key'."""
    
    value = dict[key]
    del dict[key]
    return value


def popkey(map, key, default=None):
    """Get and remove a dictionary item.

    'dict' -- The dictionary from which to get and remove the item.

    'key' -- The key of the item to get and remove.

    'default' -- The value to return if 'key' is not in 'map'.

    returns -- The value corresponding to 'key' in 'dict', or 'default'
    if none."""

    if key in map:
        result = map[key]
        del map[key]
        return result
    else:
        return default


def getHostName():
    """Return the name of this host."""

    import socket

    try:
        return socket.gethostbyname_ex(socket.gethostname())[0]
    except socket.error:
        return "(unknown)"


def isMap(object):
    """Determine whether an object satisfies the mapping protocol.

    returns -- True if 'object' satisfies the map protocol as
    specified.""" 

    # FIXME.  There probably is a more consistent way to do this.

    # If the object is an instance of the standard 'dict' class (or a
    # subclass of it), it must satisfy the mapping protocol.
    if isinstance(object, dict):
        return True
    # Check for the required methods.
    for method_name in ("get", "__getitem__", "__len__", ):
        if not hasattr(object, method_name):
            return False
    # The object checks out.
    return 1


def mkdirRecursively(path):
    """Create a directory, including parent directories if necessary."""

    parent, name = os.path.split(path)
    if parent != "" and not os.path.isdir(parent):
        mkdirRecursively(parent)
    os.mkdir(path)


def rmRecursively(path):
    """Remove the file or directory at 'path'.

    If 'path' is a directory, its contents are removed first."""

    if os.path.isdir(path):
        # Construct a list of contents of the directory.
        contents = map(lambda e, path=path: os.path.join(path, e),
                       os.listdir(path))
        # Remove them all.
        map(rmRecursively, contents)
        # Remove the directory itself.
        os.rmdir(path)
    else:
        os.unlink(path)


