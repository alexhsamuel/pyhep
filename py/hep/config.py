#----------------------------------------------------------*- python -*-
#
# module hep.config
#
# Copyright 2003 by Alex Samuel.  All rights reserved.
#
#-----------------------------------------------------------------------

import os

version = "0.8.1"

# Determine the path to the PyHEP installation we're running.
base_dir = os.path.dirname(os.path.dirname(os.path.realpath(__file__)))
data_dir = os.path.join(base_dir, "hep", "data")

