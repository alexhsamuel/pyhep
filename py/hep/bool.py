#-----------------------------------------------------------------------
#
# module hep.bool
#
# Copyright 2003 by Alex Samuel.  All rights reserved.
#
#-----------------------------------------------------------------------

"""Implementation of bool for older Python versions.

This module provides 'bool', 'True', and 'False' similar to those found
in recent versions of Python.  These symbols are defined only when
Python doesn't include its own built-in."""

#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

from   hep.ext import BoolArray

#-----------------------------------------------------------------------

if not hasattr(__builtins__, "bool"):

    False = 0

    True = 1


    class bool(object):

        def __new__(class_, value):
            if value:
                return True
            else:
                return False



