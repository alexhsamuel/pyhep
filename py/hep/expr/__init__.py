#-----------------------------------------------------------------------
#
# module hep.expr
#
# Copyright 2003 by Alex Samuel.  All rights reserved.
#
#-----------------------------------------------------------------------

"""Compiled arithmetical expressions."""

#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

from   classes import *
import compiler
from   compiler import isCompiledExpression
import expr_parser
from   hep.bool import *
import hep.ext
import inspect
import op
from   op import optimize, substitute, substituteExpressions
from   op import setTypes, setTypesFrom, setTypesFixed
import re
from   symbols import substituteConstants

#-----------------------------------------------------------------------
# functions
#-----------------------------------------------------------------------

def isExpression(object):
    """Return true if 'object' is an expression object."""

    return isCompiledExpression(object) \
           or isinstance(object, Expression)


def asExpression(arg, compile=False, types=None, types_from=None, **names):
    """Interpret 'arg' as an expression.

    For maximum convenience, this function is called by (most)
    user-visible functions which take expression arguments.

    'compile' -- If true, returns a compiled expression.

    '**names' -- Values to substitute for symbol names in the
    expression.

    returns -- If 'arg' is an expression, itself; or 'arg' converted to
    an expression.

    raises -- 'TypeError' if 'arg' is not an expression and cannot be
    converted to one."""

    # If it's already an expression, just use it.
    if isExpression(arg):
        expr = arg

    # If it's a string, parse it.
    elif isinstance(arg, str):
        try:
            expr = parse(arg)
        except expr_parser.ParseError, exception:
            # Make parse exceptions more useful by including the parsed
            # string. 
            raise expr_parser.ParseError, \
                  "%s, while parsing \"%s\"" % (exception, arg)

    # If it's a function, fish out its argument names, make symbols out
    # of these, and wrap it all in a call expression.
    elif callable(arg):
        if inspect.isfunction(arg):
            # An ordinary function.
            symbol_names = inspect.getargspec(arg)[0]
        elif inspect.ismethod(arg):
            # A class method.
            symbol_names = inspect.getargspec(arg.im_func)[0]
            # If it's a bound method, skip the first ('self') argument.
            if arg.im_self is not None:
                symbol_names = symbol_names[1 :]
        elif True:
            symbol_names = ("x", )
        else:
            # Don't know how to get the symbol names.
            raise ValueError, \
                  "don't know how to get argument names for '%r'" \
                  % arg
        # Construct the expression.
        subexprs = map(Symbol, symbol_names)
        expr = Call(Constant(arg), subexprs)

    # If it's numerical, wrap it in a constant.
    elif type(arg) in (int, float, long):
        expr = Constant(arg)

    # Don't know how to handle other things.
    else:
        raise TypeError, "argument must be an expression"

    # Substitute names.
    expr = substitute(expr, **names)
    # Set types.
    if types_from is not None:
        expr = setTypesFrom(expr, **types_from)
    if types is not None:
        expr = setTypes(expr, **types)
    # Compile it, if requested.
    if compile:
        expr = compiler.compile(expr)
    # All done.
    return expr


def parse(expression_string, **names):
    expression = expr_parser.parse(expression_string)
    expression = substitute(expression, **names)
    return expression


def compile(expression, default_type=None, **types):
    """Compile 'expression'.

    'expression' -- """

    if isCompiledExpression(expression):
        # It is already compiled.  
        return expression
    expression = asExpression(expression)
    
    if default_type is not None:
        # Set default symbol types.
        expression = op.setTypesFixed(expression, None, default_type)
    if len(types) != 0:
        # Set specific symbol types.
        expression = op.setTypes(expression, **types)

    # Do the compilation.
    expression = op.optimize(expression)
    return compiler.compile(expression)


def parseFile(lines, symbols={}, expressions=None):
    if expressions is None:
        expressions = {}

    name_regex = re.compile(r"(\w+)\s*=(\s|\Z)")

    name = None
    text = None
    def finish():
        if name is None:
            return
        expression = substituteExpressions(parse(text), expressions)
        expression = substitute(expression, **symbols)
        expressions[name] = expression

    for line in lines:
        line = line.strip()
        # Trim off comments.
        if "#" in line:
            line = line[:line.index("#")]
        # Skip blank lines.
        if line.strip() == "":
            continue

        # Is this the start of a new definition?
        match = name_regex.match(line)
        if match is not None:
            finish()
            name = match.group(1)
            text = line[match.end():]

        elif name is not None:
            text += " " + line

        else:
            raise RuntimeError, "syntax error"
        
    finish()
    return expressions


def _getSymbolNames(expr, result):
    if isinstance(expr, Symbol):
        name = expr.symbol_name
        result[name] = None
    for subexpr in expr.subexprs:
        _getSymbolNames(subexpr, result)


def getSymbolNames(expr):
    """Return a list of names of symbols in 'expr'."""
    
    result = {}
    _getSymbolNames(expr, result)
    return result.keys()
