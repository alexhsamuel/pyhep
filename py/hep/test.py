#-----------------------------------------------------------------------
#
# module hep.test
#
# Copyright 2003 by Alex Samuel.  All rights reserved.
#
#-----------------------------------------------------------------------

"""Regression testing support."""

#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

from   hep.bool import *
import sys

#-----------------------------------------------------------------------
# exceptions
#-----------------------------------------------------------------------

class TestFailure(Exception):
    pass



#-----------------------------------------------------------------------
# functions
#-----------------------------------------------------------------------

def assert_(expression):
    if not expression:
        raise TestFailure, "assertion failed"


def compare(value, expected, precision=None):
    if type(value) in (list, tuple) and type(expected) in (list, tuple):
        if len(value) != len(expected):
            result = False
        else:
            try:
                result = reduce(
                    lambda r1, r2: r1 and r2,
                    map(lambda (v, e): compare(v, e, precision),
                        zip(value, expected)),
                    True)
            except TestFailure:
                result = False

    elif precision is not None:
        result = (abs(value - expected) <= precision)

    else:
        result = (value == expected)

    if result:
        return True
    else:
        raise TestFailure, \
              "expected %s; got %s" % (repr(expected), repr(value))


def compareSequence(value, expected, ordered=True):
    value = list(value)
    expected = list(expected)
    if not ordered:
        value.sort()
        expected.sort()
    return compare(value, expected)


