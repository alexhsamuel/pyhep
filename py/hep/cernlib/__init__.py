#-----------------------------------------------------------------------
#
# module hep.cernlib
#
# Copyright 2003 by Alex Samuel.  All rights reserved.
#
#-----------------------------------------------------------------------

"""CERNLIB library interface."""

#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

import ext
from   hep import popkey
from   hep.bool import *

#-----------------------------------------------------------------------
# functions
#-----------------------------------------------------------------------

def integrate(function, *region, **kw_args):
    """Numerically integrate 'function'.

    Performs multidimensional numerical integration of 'function' over a
    rectangular integration 'region'.

    'function' -- A callable or expression.

    '*region' -- One-dimensional ranges specifying a rectangular
    integration region.  Each range is a tuple of the form '(var_name,
    lo, hi)', where 'var_name' is the name of the integration variable
    and 'lo' and 'hi' are 'float' values specifying the interval to
    integrate over.

    'accuracy' -- Specify the desired accuracy with this keyword
    argument.

    returns -- The integral value."""

    # If 'function' isn't already callable, treat it as an expression. 
    import hep.expr
    if not callable(function):
        function = hep.expr.asExpression(function)

    if hep.expr.isExpression(function) \
       and not hep.expr.isCompiledExpression(function):
        function = hep.expr.setTypesFixed(function, None, float)
        function = hep.expr.compile(function)

    accuracy = float(popkey(kw_args, "accuracy", 1e-8))
    if len(kw_args) > 0:
        raise TypeError, \
              "'integrate' got an unexpected keyword argument '%s'" \
              % kw_args.keys()[0]

    # Perform the integration.
    return ext.integrate(function, region, accuracy)


