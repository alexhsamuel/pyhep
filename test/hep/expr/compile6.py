#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

import hep.expr
from   hep.test import compare

#-----------------------------------------------------------------------
# main script
#-----------------------------------------------------------------------

i1 = 42
i2 = -4
f1 = 0.5
f2 = 3.0
symbols = { "i1": i1, "i2": i2, "f1": f1, "f2": f2 }

tests = [
    ("i1", i1),
    ("(i1)", (i1)),
    ("(i1, )", (i1, )),
    ("(i1, i2)", (i1, i2)),
    ("(i1, i2, f1, f2)", (i1, i2, f1, f2)),
    ("(i1 + 2, f1 / f2)", (i1 + 2, f1 / f2)),
    ]

for expression, expected in tests:
    compiled = hep.expr.asExpression(
        expression, types_from=symbols, compile=True)
    assert hep.expr.isCompiledExpression(compiled)
    value = compiled.evaluate(symbols)
    compare(value, expected)
    compare(type(value), type(expected))
