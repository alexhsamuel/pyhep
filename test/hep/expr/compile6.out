Traceback (most recent call last):
  File "compile6.py", line 28, in ?
    compiled = hep.expr.asExpression(
  File "/home/samuel/dev/pyhep/py/hep/expr/__init__.py", line 67, in asExpression
    raise expr_parser.ParseError, \
hep.expr.expr_parser.ParseError: unknown symbol testlist_gexp, while parsing "(i1)"
